llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp

```diff
@@ -1270,13 +1270,22 @@ AMDGPULegalizerInfo::AMDGPULegalizerInfo(const GCNSubtarget &ST_,
     .custom();
 
   // The 64-bit versions produce 32-bit results, but only on the SALU.
-  getActionDefinitionsBuilder({G_CTLZ_ZERO_UNDEF, G_CTTZ_ZERO_UNDEF})
-    .legalFor({{S32, S32}, {S32, S64}})
-    .clampScalar(0, S32, S32)
-    .clampScalar(1, S32, S64)
-    .scalarize(0)
-    .widenScalarToNextPow2(0, 32)
-    .widenScalarToNextPow2(1, 32);
+  getActionDefinitionsBuilder(G_CTLZ_ZERO_UNDEF)
+      .legalFor({{S32, S32}, {S32, S64}})
+      .customFor({{S32, S8}, {S32, S16}})
```

Comment:
This shouldn't really need to specially handle illegal typed sources. I would expect to be able to recover the promotion that would happen anyway with KnownBits. If you're going to special case the illegal types, might as well handle arbitrary bit width sources here

Reply:
It doesn't matter how many bits it was originally. You're shifting away the high bits, it doesn't need to be restricted to 8 or 16

Reply:
I think that would give us the wrong result. The current lowering subtracts the number of extended bits from the result of CTLZ, so the number of leading zeroes in the original value needs to be maintained.